<?xml version="1.0"?>
<!--
  Copyright (C) 2022 LEIDOS.

  Licensed under the Apache License, Version 2.0 (the "License"); you may not
  use this file except in compliance with the License. You may obtain a copy of
  the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
  License for the specific language governing permissions and limitations under
  the License.
-->

<!--
Example launch file for the CARLA Sensors Integration node
Loads parameters and configures logging for the node

 -->

 <launch>

    <arg name='role_name' default='ego_vehicle'/>
    <arg name='object_detection_stream' default="false"/>
    <arg name="localization_stream" default="false"/>
    <arg name="carla_lidar_stream" default="true"/>
    <arg name="carla_camera_stream" default="true"/>
    <arg name="carla_gnss_stream" default="true"/>

    <remap from="points_raw" to="/hardware_interface/lidar/points_raw"/>
    <remap from="image_raw" to="/hardware_interface/camera/image_raw"/>
    <remap from="image_color" to="/hardware_interface/camera/image_color"/>
    <remap from="image_rect" to="/hardware_interface/camera/image_rects"/>
    <remap from="camera_info" to="/hardware_interface/camera/camera_info"/>
    <remap from="gnss_fix_fused" to="/hardware_interface/gnss_fix_fused"/>

    <!--
      # localization #
      Extract the pose, twist from carla odometry.
    -->
    <group if="$(arg localization_stream)">
      <node pkg='carma_carla_bridge' type='carla_to_carma_localization' name='carla_to_carma_localization' output='screen'>
        <param name='role_name' type="str" value="$(arg role_name)"/>
      </node>
    </group>

    <!--
      # external objects
      Extract the external objects from carla ObjectArray
    -->
    <group if="$(arg object_detection_stream)">
      <node pkg='carma_carla_bridge' type='carla_to_carma_external_objects' name='carla_to_carma_external_objects' output='screen'>
        <param name='role_name' type="str" value="$(arg role_name)"/>
        <param name="object_detection_stream" value="true" type="bool"/>
      </node>
    </group>

    <node pkg="carla_sensors_integration" type="carla_sensors_integration" name="carla_sensors_integration" output='screen'>
      <param name='role_name' type="str" value="$(arg role_name)" />
      <param name='object_detection_stream' type="bool" value="$(arg object_detection_stream)" />
      <param name='localization_stream' type="bool" value="$(arg localization_stream)" />
      <param name='carla_lidar_stream' type="bool" value="$(arg carla_lidar_stream)" />
      <param name='carla_camera_stream' type="bool" value="$(arg carla_camera_stream)" />
      <param name='carla_gnss_stream' type="bool" value="$(arg carla_gnss_stream)" />
    </node>

 </launch>
